cassiopeia/tools/branch_length_estimator.py

```diff
@@ -0,0 +1,325 @@
+    https://docs.mosek.com/modeling-cookbook/expo.html
+    Because it is a convex program, it can be readily solved.
+
+    Args:
+        minimum_branch_length: Estimated branch lengths will be constrained to
+            have at least this lenght.
+        l2_regularization: Consecutive branches will be regularized to have
+            similar length via an L2 penalty whose weight is given by
+            l2_regularization.
+        verbose: Verbosity level.
+
+    Attributes:
+        log_likelihood: The log-likelihood of the training data under the
+            estimated model.
+        log_loss: The log-loss of the training data under the estimated model.
+            This is the log likhelihood plus the regularization terms.
+    """
+
+    def __init__(
+        self,
+        minimum_branch_length: float = 0,
+        l2_regularization: float = 0,
+        verbose: bool = False,
+    ):
+        self.minimum_branch_length = minimum_branch_length
+        self.l2_regularization = l2_regularization
+        self.verbose = verbose
+
+    def estimate_branch_lengths(self, tree: Tree) -> None:
+        r"""
+        See base class. The only caveat is that this method raises if it fails
+        to solve the underlying optimization problem for any reason.
+
+        Raises:
+            cp.error.SolverError
+        """
+        # Extract parameters
+        minimum_branch_length = self.minimum_branch_length
+        l2_regularization = self.l2_regularization
+        verbose = self.verbose
+
+        # # Wrap the networkx DiGraph for goodies.
+        # tree = Tree(tree)
+
+        # # # # # Create variables of the optimization problem # # # # #
+        r_X_t_variables = dict(
+            [
+                (node_id, cp.Variable(name=f"r_X_t_{node_id}"))
+                for node_id in tree.nodes()
+            ]
+        )
+        time_increases_constraints = [
+            r_X_t_variables[parent]
+            >= r_X_t_variables[child] + minimum_branch_length
+            for (parent, child) in tree.edges()
+        ]
+        leaves_have_age_0_constraints = [
+            r_X_t_variables[leaf] == 0 for leaf in tree.leaves()
+        ]
+        non_negative_r_X_t_constraints = [
+            r_X_t >= 0 for r_X_t in r_X_t_variables.values()
+        ]
+        all_constraints = (
+            time_increases_constraints
+            + leaves_have_age_0_constraints
+            + non_negative_r_X_t_constraints
+        )
+
+        # # # # # Compute the log-likelihood # # # # #
+        log_likelihood = 0
+
+        # Because all rates are equal, the number of cuts in each node is a
+        # sufficient statistic. This makes the solver WAY faster!
+        for (parent, child) in tree.edges():
+            edge_length = r_X_t_variables[parent] - r_X_t_variables[child]
+            # TODO: hardcoded '0' here...
+            zeros_parent = tree.get_state(parent).count("0")
+            zeros_child = tree.get_state(child).count("0")
```

Comment:
Yes, as per the docstring: `This estimator requires that the ancestral states are provided.` . The estimator doesn't sanity-check the input tree currently, but we can do that. (In general, there's a bunch of sanity checks that could be added in many places and aren't there because it wasn't my focus when coding.)

As a note, the estimator currently does not have the responsibility of imputing ancestral states if they are not provided.

Reply:
Totally makes sense! When this code is deployed, let's make sure that an error is thrown if ancestral states are not given.